third_party/nvfuser/csrc/multidevice/aggregate_dag.cpp

```diff
@@ -0,0 +1,326 @@
+                    && unaryPredicate(aVal);
+            });
+    return aVals_filtered;
+  }
+
+  template<typename containerType>
+  containerType FilterAggregateValContainer(
+                containerType aVals,
+                Val* val = nullptr,
+                ClusterPtr cluster = nullptr)
+  {
+    return FilterAggregateValContainer(aVals, [](auto v){return true;}, val, cluster);
+  }
+
+  const std::vector<Val*>& AggregateDag::MGFusionInputs() const {
+    return MGFusion_->inputs();
+  }
+
+  const std::vector<Val*>& AggregateDag::MGFusionOutputs() const {
+    return MGFusion_->outputs();
+  }
+
+  // Creates the AggregateVals IR in the AggregateDag
+  // An AggregateVal is created for each input/output of each Cluster of the MultiClusterFusion
+  // If a Val is an I/O of the original MultiClusterFusion, the correspondings
+  // AggregateVal are also I/O of the AggregateDag
+  void AggregateDag::buildAVals() {
+    for (auto cluster : MGFusion_->clusters()) {
+      for (auto output_val : cluster->outputs().vector()) {
+        auto av = IrBuilder::create<AggregateVal>(this->as<IrContainer>(), output_val, cluster);
+        if (std::count(MGFusion_->outputs().begin(), 
+                       MGFusion_->outputs().end(), output_val)) {
+            addOutput(av);
+        }
+      }
+      for (auto input_val : cluster->inputs().vector()) {
+        auto av = IrBuilder::create<AggregateVal>(this->as<IrContainer>(), input_val, cluster);
+        if (std::count(MGFusion_->inputs().begin(), 
+                       MGFusion_->inputs().end(), input_val)) {
+            addInput(av);
+        }
+      }
+    }
+  }
+
+  // Build the AggregateExpr IR of the AggregateDag
+  // An AggregateExpr is created for each Cluster of the MultiClusterFusion
+  void AggregateDag::buildAExpr() {
+    for (auto cluster : MGFusion_->clusters()) {
+
+    // the inputs (resp. outputs) of AggregateExpr are the 
+    // cluster's consumed (resp. produced) vals
+      auto output_container = FilterAggregateValContainer(vals_, 
+                                      [](auto val){return val->isValOutputOfCluster();},
+                                      nullptr, cluster);
+      auto input_container = FilterAggregateValContainer(vals_, 
+                                      [](auto val){return val->isValInputOfCluster();},
+                                      nullptr, cluster);
+
+      auto aExpr = IrBuilder::create<AggregateExpr>(this->as<IrContainer>(), cluster);
+
+      for (auto v: output_container){
+        v->setDefinition(aExpr);
+      }
+    }
+  }
+
+  // Build the SendRecv IR of the AggregateDag
+  // A SendRecv is created for each pair of AggregateVals which Vals match
+  // and which are not inputs/outputs of the AggregateDag
+  void AggregateDag::buildSendRecv() {
+    //select all AggregateVal that are not global I/O of the AggregateDag
+    decltype(vals_) internal_aVals;
+    std::copy_if(vals_.begin(),
```

Comment:
Any specific reason not to use the filter routine?

Reply:
you're right, using the routine is nicer